mne/preprocessing/xdawn.py

```diff
@@ -0,0 +1,412 @@
+            self.signal_cov_ = signal_cov
+        else:
+            raise ValueError('signal_cov must be None, a covariance instance '
+                             'or a ndarray')
+
+        # estimates evoked covariance
+        self.evokeds_cov_ = {}
+        if correct_overlap:
+            if epochs.baseline is not None:
+                raise ValueError('Baseline correction must be None if overlap '
+                                 'correction activated')
+            evo, to = least_square_evoked(epochs, return_toeplitz=True)
+        else:
+            evo = {eid: epochs[eid].average() for eid in epochs.event_id}
+            to = {eid: 1.0 / len(epochs[eid]) for eid in epochs.event_id}
+        self.evokeds_ = evo
+
+        for eid in epochs.event_id:
+            # FIXME use mne covariance estimator
+            self.evokeds_cov_[eid] = np.cov(np.dot(evo[eid].data, to[eid]))
+
+        # estimates spatial filters
+        for eid in epochs.event_id:
+
+            if self.signal_cov_.shape != self.evokeds_cov_[eid].shape:
+                raise ValueError('Size of signal cov must be the same as the'
+                                 ' number of channels in epochs')
+
+            evals, evecs = linalg.eigh(self.evokeds_cov_[eid],
+                                       self.signal_cov_)
+            evecs = evecs[:, np.argsort(evals)[::-1]]  # sort eigenvectors
+            evecs /= np.apply_along_axis(np.linalg.norm, 0, evecs)
+
+            self.filters_[eid] = evecs
+            self.patterns_[eid] = linalg.inv(evecs.T)
+
+            # Create and add projectors
+            ch_names = self.evokeds_[eid].ch_names
+            projs = []
+            for j in range(self.n_components):
+                proj_data = dict(col_names=ch_names, row_names=None,
+                                 data=evecs[:, j].T, nrow=1,
+                                 ncol=len(ch_names))
+                projs.append(Projection(active=True, data=proj_data,
+                             desc="%s Xdawn #%d" % (eid, j)))
+
+            self.projs_[eid] = projs
+
+        # store some values
+        self.ch_names = epochs.ch_names
+        self.exclude = range(self.n_components, len(self.ch_names))
+        self.event_id = epochs.event_id
+        return self
+
+    def apply(self, inst, event_id=None, include=None, exclude=None):
+        """Remove selected components from the signal.
+
+        Given the unmixing matrix, transform data,
+        zero out components, and inverse transform the data.
+        This procedure will reconstruct M/EEG signals from which
+        the dynamics described by the excluded components is subtracted.
+
+        Parameters
+        ----------
+        inst : instance of Raw, Epochs or Evoked
+            The data to be processed.
+        event_id : dict or None
+            The kind of event to apply. if none, a dict of inst will be return
+            one for each type of event xdawn has been fitted.
```

Comment:
does this apply just to Epochs? If so I would remove this option and just pass Epochs for the event of interest